obspy/kinemetrics/evt.py

```diff
@@ -0,0 +1,411 @@
+            file_pointer = open(filename_or_object, "rb")
+
+        self.e_tag.read(file_pointer)
+        endian = self.e_tag.endian
+        self.e_header.unsetdico()
+        self.e_header.read(file_pointer, self.e_tag.length, endian)
+
+        self.data = np.ndarray([self.e_header.nchannels, 0])
+
+        while True:
+            try:
+                self.e_tag.read(file_pointer)
+                retparam = self.e_frame.read(file_pointer,
+                                             self.e_tag.length, endian)
+                if self.samplingrate == 0:
+                    self.samplingrate = retparam[0]
+                elif self.samplingrate != retparam[0]:
+                    raise EvtBadHeaderError("Sampling rate not constant")
+                datal = self.e_data.read(file_pointer,
+                                         self.e_tag.datalength,
+                                         endian, retparam)
+                npdata = np.array(datal)
+                self.data = np.hstack((self.data, npdata))  # append data
+            except EvtEOFError:
+                break
+        if self.e_frame.count() != self.e_header.duration:
+            raise EvtBadDataError("Bad number of blocks")
+
+        if not raw:
+            self.calibration()
+        if is_fileobject is False:
+            file_pointer.close()
+
+        traces = []
+        for i in range(self.e_header.nchannels):
+            cur_trace = Trace(data=self.data[i])
+            cur_trace.stats.channel = str(i)
+            cur_trace.stats.station = self.e_header.stnid.replace(b"\x00",
+                                                                  b"").decode()
+            cur_trace.stats.sampling_rate = float(self.samplingrate)
+            cur_trace.stats.starttime = self.e_header.starttime
+            cur_trace.stats.kinemetrics_evt = self.e_header.makeobspydico(i)
+            traces.append(cur_trace)
+
+        return Stream(traces=traces)
+
+
+class EvtData(object):
+    """
+    Class to manage data in Evt file
+    """
+    def read(self, file_p, length, endian, param):
+        """
+        read data from file_p
+
+        :param file_p: file pointer
+        :param length: length to be read
+        :param endian: endian type in datafile
+        :type param: list
+        :param param: sampling rate,sample size, block time, channels
+        :rtype: list of list
+        :return: list of data
+        """
+        buff = file_p.read(length)
+        samplerate = param[0]
+        numbyte = param[1]
+        numchan = param[3]
+        num = (samplerate / 10) * numbyte * numchan
+        data = [[] for _ in range(numchan)]
+        if length != num:
+            raise EvtBadDataError("Bad data length")
+        for j in range(samplerate // 10):
+            for k in range(numchan):
+                i = (j * numchan) + k
+                if numbyte == 2:
+                    val = unpack(b">i", buff[i * 2:(i * 2) + 2] + b'\0\0')[0] \
```

Comment:
See my previous mail.
We have no 2 or 4 bytes files. Impossible to test.
The original code is based on kinemetrics C code and should work.
The code modified by Eliott Sales de Andrade : i am not sure, because 
the code is not // to 3 bytes code.
Perhaps ask to kinemetrics to have a 2bytes and 4bytes example code ?

Henri


On 26/01/2015 08:23, Thomas Lecocq wrote:
>
> In obspy/kinemetrics/evt.py 
> <https://github.com/obspy/obspy/pull/781#discussion_r23514972>:
>
> > +        :rtype: list of list
> > +        :return: list of data
> > +        """
> > +        buff = file_p.read(length)
> > +        samplerate = param[0]
> > +        numbyte = param[1]
> > +        numchan = param[3]
> > +        num = (samplerate / 10) * numbyte * numchan
> > +        data = [[] for _ in range(numchan)]
> > +        if length != num:
> > +            raise EvtBadDataError("Bad data length")
> > +        for j in range(samplerate // 10):
> > +            for k in range(numchan):
> > +                i = (j * numchan) + k
> > +                if numbyte == 2:
> > +                    val = unpack(b">i", buff[i * 2:(i * 2) + 2] + b'\0\0')[0] \
>
> @vcoeur61 <https://github.com/vcoeur61> : can you check this ?
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/obspy/obspy/pull/781/files#r23514972>.
>